TRPC 11

vafanassieff opened the pull request:
Hey there, thanks for this package.


Best

## 🎯 Changes

- changed repo structure to a monorepo with `pnpm`
- use trpc@next since it's a release candidate
- new handler generation (from @louneskmt)
- use `links` like a normal TRPC client (from @louneskmt)
- mock subscription with `msw@next` (from @louneskmt)
- use vitest as test runner
- add tests for node and react

## 🔎  Todo

- add documentation for new API with example
- fix workflows

## ✅ Checklist

- [x] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

l0gicgate commented on packages/msw-trpc/package.json:
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "msw-trpc",
+  "version": "2.0.0-beta.1",
+  "description": "Trpc API for Mock Service Worker (MSW).",
```
```suggestion
  "description": "tRPC API for Mock Service Worker (MSW).",

l0gicgate commented on packages/msw-trpc/src/create.ts:
```diff
@@ -0,0 +1,44 @@
+  const { links, transformer } = config
+
+  // @ts-expect-error any
+  const createUntypedTRPCMsw = (pathParts: string[] = []) => {
```
Just a nit, but I would call it `pathSegments`. 

Parts feels odd

maloguertin replied:
https://developer.mozilla.org/en-US/docs/Learn/Common_questions/Web_mechanics/What_is_a_URL parts seems to be a term that is used when talking about url's so I'll resolve for now!

l0gicgate commented on packages/test-react/package.json:
```diff
@@ -0,0 +1,25 @@
+    "@trpc/server": "11.0.0-rc.403",
+    "@types/react": "18.3.3",
+    "@types/react-dom": "18.3.0",
+    "jsdom": "24.1.0",
```
happy-dom is faster than jsdom, albeit it is lacking some of the web APIs, one should consider trying to see if it's adequate for this package

https://github.com/capricorn86/happy-dom

l0gicgate reviewed (approved):
@vafanassieff I've reviewed the entire PR, it looks great. I am not a core contributor though but I am looking forward to seeing this merged as we are using tRPC v11 and would like to use this package for testing.@vafanassieff I've reviewed the entire PR, it looks great. I am not a core contributor though but I am looking forward to seeing this merged as we are using tRPC v11 and would like to use this package for testing.

vafanassieff commented:
> @vafanassieff I've reviewed the entire PR, it looks great. I am not a core contributor though but I am looking forward to seeing this merged as we are using tRPC v11 and would like to use this package for testing.

Thanks !
The PR is just missing an updated documentation, we changed the API and added the Websocket API.
And `msw` websocket is still in beta.

kenny-grubb commented on packages/msw-trpc/src/handler.ts:
```diff
@@ -0,0 +1,285 @@
+}
+
+const getMutationInput = async (req: Request, transformer: TRPCCombinedDataTransformer) => {
+  const body = await req.json()
```
With TRPC 11, it is possible that a request will contain FormData (ex: for file uploads), and this code will throw an error.

maloguertin replied:
Thanks for the input, as this code is not new I think we could put a new milestone to support different input types.

maloguertin replied:
https://github.com/maloguertin/msw-trpc/issues/43

randseay commented:
Our team is excited to see support for tRPC v11, thank you! Is there a chance this might merge soon?

vafanassieff commented:
> @vafanassieff could you add the documentation and then we could go forward with this PR? Thanks!

Hey I currently don't have much time to write the docs, if someone want to chat with me and @louneskmt to have some guidance and help us!

jasonsilvers-dispel commented:
@vafanassieff  @louneskmt I would be happy to help with the documentation. The github action to publish might also need to be updated as well. I ran into a view issues with it when creating a temporary fork.

vafanassieff commented:
Hey, I removed the websocket part since it added too much overhead for almost nothing.
We can add this later, I think HTTP link is enough for a first release.
The doc is updated, it's pretty straightforward, use `msw-trpc` like a regular tRPC client.

blackmax1886 commented:
I really appreciate this library & this PR.
I wrote the following test code with vitest & RTL based on this pull request, but something weird was happened

# trpc.tsx
```typescript
export const trpcMsw = createTRPCMsw<AppRouter>({
  transformer: { input: SuperJSON, output: SuperJSON },
  links: [httpLink({ url: "http://localhost:3000/api/trpc" })],
});

const queryClient = new QueryClient();
const trpc = createTRPCReact<AppRouter>();
const client = trpc.createClient({
  links: [
    TRPChttpLink({
      url: "http://localhost:3000/api/trpc",
      transformer: SuperJSON,
    }),
  ],
});

const MockedTRPCReactProvider = (props: { children: React.ReactNode }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <trpc.Provider client={client} queryClient={queryClient}>
        {props.children}
      </trpc.Provider>
    </QueryClientProvider>
  );
};

export const renderWithTRPC = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => {
  return render(ui, {
    wrapper: (props) => <MockedTRPCReactProvider {...props} />,
    ...options,
  });
};

```
# posts.spec.tsx
```typescript
const server = setupServer();
beforeAll(() => {
  server.listen();
});
afterEach(() => {
  server.resetHandlers();
});

afterAll(() => server.close());

test("should display `No posts yet` with no posts", async () => {
  server.use(trpcMsw.post.all.query(() => []));

  renderWithTRPC(<PostList />);
  await waitFor(() => {
    expect(screen.getByText("No posts yet")).toBeInTheDocument();
  });
  server.close();
});

test("should display posts", async () => {
  const posts = [
    {
      id: "sample-id-1",
      title: "title1",
      content: "content1",
      createdAt: new Date(),
      updatedAt: new Date(),
    }
  ];
  server.use(trpcMsw.post.all.query(() => posts)); // setting hanlder works

  renderWithTRPC(<PostList />); // but the component is rendered with no posts

// fail
  await waitFor(() => {
    expect(screen.getAllByRole("heading")).toHaveLength(1);
  });
  expect(screen.getByRole("heading", { name: "title1" })).toBeInTheDocument();
  expect(screen.getByText("content1")).toBeInTheDocument();

  server.close();
});
```

If I use server.use() multiple times on the same file for the same tRPC endpoint, the second tRPC responses are fixed at the value mocked the first time, even though I'm running server.resetHandlers() in afterEach.
Reversing the order would cause the first test to pass and the second to fail.

It was confirmed by listingHandlers() that the handlers were reset and set correctly for each test.

It was also confirmed by screen.debug() etc. that the DOM is rendered with the return value of the first mock.

cleanup from @testing-library in afterEach did not affect.

It seems that renderWithTRPC is not working as intended. If something wrong, pls let me know.
thanks.

- msw: 2.4.6
- vitest: 2.1.2
- trpc: 11.0.0-rc.477

rinarakaki commented:
@vafanassieff could you please resolved the conflicts?

rinarakaki commented:
@maloguertin could you please take a look at the doc update?

rinarakaki commented on README.md:
```diff
@@ -41,143 +44,45 @@ export const trpcMsw = createTRPCMsw<AppRouter>() /* 👈 */
 ## Config
 
-`createTRPCMsw` accepts a 2nd argument:
+You need to pass a `httpLink` to the `createTRPCMsw` function like you would do with the tRPC client.
```
This description must come before the `**2. build your trpcMsw with createTRPCMsw.**` in the `## Usage` section  above since the `config` argument of `createTRPCMsw` is now non-optional.

danilofuchs commented:
@maloguertin msw has released WebSocket support in version 2.6.0

https://github.com/mswjs/msw/releases/tag/v2.6.0

I love this package and have been using it extensively with @vafanassieff fork

Is there anything else the community is able to provide in this PR to unblock it and start implementing WebSocket, FormData and other goodies?

vafanassieff commented:
> @maloguertin msw has released WebSocket support in version 2.6.0
> 
> https://github.com/mswjs/msw/releases/tag/v2.6.0
> 
> I love this package and have been using it extensively with @vafanassieff fork
> 
> Is there anything else the community is able to provide in this PR to unblock it and start implementing WebSocket, FormData and other goodies?

If you look in the previous commit, we added a way to use websocket, but it was kinda cluncky and out of the scope of this PR.

(BTW we moved to SSE instead of websocket for subscriptions, it's much easier)

Mathieu-R commented:
Hi !
I'm currently using it in a project and this fork works just fine (at least for now) with TRPC v11 and MSW v2.

Any plan to merge ? @vafanassieff  Did you publish your fork on npm under a prefix ?

vafanassieff commented:
> Hi ! I'm currently using it in a project and this fork works just fine (at least for now) with TRPC v11 and MSW v2.
> 
> Any plan to merge ? @vafanassieff Did you publish your fork on npm under a prefix ?

I'm using `"@vafanassieff/msw-trpc": "0.0.1"` meanwhile.

Alletkla commented:
Hello 👋🏽 
@vafanassieff  
Do you know what is blocking this PR from being merged? - I would really appreciate it as it blocks update to trpc11 for me and my team.
Since websocket support has been removed from this PR, 
the only outstanding issue I see is in the documentation, where the explanation for createTRPCMsw/config should be moved up.

The only conflict with main is also in the documentation. I can solve this.
Although unfortunately I probably can't help with in depth programming due to lack of experience.

vafanassieff commented:
@Alletkla 

I have yet to find the time to write the doc, if you are willing to do it lfg !

balloman commented:
Just as a heads up, this doesn't seem to support post overrides (it always assumes a get request for a query)

louneskmt commented:
@balloman This should do the trick ba00546.

vafanassieff commented:
I updated the PR to use tRPC 11.

@Alletkla if you still want to update the config lmk :)

vatheara commented:
this is really helpful in storybook please merge

l0gicgate commented:
At this point, it appears the maintainer of this repo does not care. I ended up forking the repo and merging this PR into my fork instead and published to NPM to use on my projects, I suggests others do the same. This PR has been open for a year now.
